### PR TITLE
[KYUUBI #2644] Add etcd discovery client for HA

### PR DESCRIPTION
**Pull request by hddong**
### _Why are the changes needed?_

Add etcd discovery client for HA

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by hddong** [comment]
cc @yaooqinn @pan3793

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdDiscoveryClient.scala [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+  private def createPersistentNode(
+      conf: KyuubiConf,
+      namespace: String,
+      instance: String,
+      version: Option[String] = None,
+      external: Boolean = false): ServiceNode = {
+    val ns = DiscoveryPaths.makePath(null, namespace)
+    create(ns, "PERSISTENT")
+
+    val session = conf.get(HA_ZK_ENGINE_REF_ID)
```
Comment (zwangsheng) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdDiscoveryClient.scala:271]: Maybe we should rename this configuration as it is used in both ETCD and ZK

**Comment by hddong** [comment]
@yaooqinn @pan3793 : Please have a review when free.

File: dev/dependencyList [hunk 2]
```diff
@@ -59,6 +68,7 @@ jakarta.validation-api/2.0.2//jakarta.validation-api-2.0.2.jar
 jakarta.ws.rs-api/2.1.6//jakarta.ws.rs-api-2.1.6.jar
 jakarta.xml.bind-api/2.3.2//jakarta.xml.bind-api-2.3.2.jar
 javassist/3.25.0-GA//javassist-3.25.0-GA.jar
+javax.annotation-api/1.3.2//javax.annotation-api-1.3.2.jar
```
Comment (pan3793) [inline dev/dependencyList:71]: We don't need this, `jakarta.annotation-api` is a drop-in replacement

File: dev/dependencyList [hunk 3]
```diff
@@ -87,9 +101,15 @@ metrics-json/4.2.8//metrics-json-4.2.8.jar
 metrics-jvm/4.2.8//metrics-jvm-4.2.8.jar
 netty-all/4.1.73.Final//netty-all-4.1.73.Final.jar
 netty-buffer/4.1.73.Final//netty-buffer-4.1.73.Final.jar
+netty-codec-dns/4.1.74.Final//netty-codec-dns-4.1.74.Final.jar
```
Comment (pan3793) [inline dev/dependencyList:104]: gRPC uses a shaded netty, why do we need those?

File: externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider [hunk 4]
```diff
@@ -0,0 +1,3 @@
+io.grpc.internal.PickFirstLoadBalancerProvider
+io.grpc.util.SecretRoundRobinLoadBalancerProvider$Provider
+io.grpc.grpclb.GrpclbLoadBalancerProvider
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:3]: What doesn't relocate them?

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala [hunk 5]
```diff
@@ -43,7 +43,7 @@ trait DiscoveryClient extends Logging {
   /**
    * Get the stored data under path.
    */
-  def getData(path: String): Array[Byte]
+  def getData(path: String): String
```
Comment (pan3793) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala:46]: We'd better avoid changing the API as much as possible.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdDiscoveryClient.scala [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+      external: Boolean = false): ServiceNode = {
+    val ns = DiscoveryPaths.makePath(null, namespace)
+    create(ns, "PERSISTENT")
+
+    val session = conf.get(HA_ENGINE_REF_ID)
+      .map(refId => s"refId=$refId;").getOrElse("")
+    val pathPrefix = DiscoveryPaths.makePath(
+      namespace,
+      s"serviceUri=$instance;version=${version.getOrElse(KYUUBI_VERSION)};${session}sequence=")
+    val znodeData = instance
```
Comment (pan3793) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdDiscoveryClient.scala:276]: znode?

File: externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider [hunk 7]
```diff
@@ -0,0 +1,3 @@
+io.grpc.internal.PickFirstLoadBalancerProvider
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:1]: license can be added here

File: externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.NameResolverProvider [hunk 8]
```diff
@@ -0,0 +1,4 @@
+io.etcd.jetcd.resolver.DnsSrvResolverProvider
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.NameResolverProvider:1]: ditto

File: pom.xml [hunk 9]
```diff
@@ -110,6 +110,7 @@
         <commons-lang.version>2.6</commons-lang.version>
         <commons-lang3.version>3.10</commons-lang3.version>
         <curator.version>2.12.0</curator.version>
+        <etcd.version>0.7.0</etcd.version>
```
Comment (yaooqinn) [inline pom.xml:113]: 0.7.1？

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdUtils.scala [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+package org.apache.kyuubi.ha.client.etcd
+
+import java.nio.charset.StandardCharsets.UTF_8
+
+import scala.language.implicitConversions
+
+import io.etcd.jetcd.ByteSequence
+
+object EtcdUtils {
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdUtils.scala:26]: seems not very necessary to have implicit here

File: dev/dependencyList (hunk 2 above)
Comment (hddong) [reply dev/dependencyList]: Removed

File: dev/dependencyList (hunk 3 above)
Comment (hddong) [reply dev/dependencyList:105]: > gRPC uses a shaded netty, why do we need those netty stuff?

These netty jars are a dependency of jetcd, had try to use netty-shaded, but got `ClassNoDefException`. Classes in  `netty-shaded` have prefix `io.grpc`, jetcd can not recognized. Any suggestions?

File: externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider (hunk 4 above)
Comment (hddong) [reply externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:20]: > What doesn't relocate them?

Had do this, and got
```scala
Caused by: java.lang.IllegalArgumentException: cannot find a NameResolver for ip:///100.71.36.42:2389
        at org.apache.kyuubi.shade.io.grpc.internal.ManagedChannelImpl.getNameResolver(ManagedChannelImpl.java:776)
        at org.apache.kyuubi.shade.io.grpc.internal.ManagedChannelImpl.getNameResolver(ManagedChannelImpl.java:785)
        at org.apache.kyuubi.shade.io.grpc.internal.ManagedChannelImpl.<init>(ManagedChannelImpl.java:665)
        at org.apache.kyuubi.shade.io.grpc.internal.ManagedChannelImplBuilder.build(ManagedChannelImplBuilder.java:630)
        at org.apache.kyuubi.shade.io.grpc.internal.AbstractManagedChannelImplBuilder.build(AbstractManagedChannelImplBuilder.java:297)
        at org.apache.kyuubi.shade.io.vertx.grpc.VertxChannelBuilder.build(VertxChannelBuilder.java:302)
        at io.etcd.jetcd.impl.ClientConnectionManager.getChannel(ClientConnectionManager.java:89)
```
Although we provide `IPResolverProvider` in `nameresolverprovider`, `NameResolver` can not load `IPResolverProvider`, we can relocate them after a solution is found.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdDiscoveryClient.scala (hunk 6 above)
Comment (hddong) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdDiscoveryClient.scala]: done

File: externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider (hunk 7 above)
Comment (hddong) [reply externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:18]: > license can be added here

Done

File: externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.NameResolverProvider (hunk 8 above)
Comment (hddong) [reply externals/kyuubi-spark-sql-engine/src/main/resources/META-INF/services/io.grpc.NameResolverProvider:18]: Done

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdUtils.scala (hunk 10 above)
Comment (hddong) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/etcd/EtcdUtils.scala]: Done

File: pom.xml (hunk 9 above)
Comment (hddong) [reply pom.xml]: Done

File: dev/dependencyList (hunk 3 above)
Comment (pan3793) [reply dev/dependencyList:105]: https://github.com/etcd-io/jetcd/pull/1068 seems going to provide an all-in-one client jar then we can get rid of those dependencies.

**Comment by hddong** [comment]
@yaooqinn @pan3793 @zwangsheng : Took some time to fix the tests, can review this again.

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/etcd/EtcdDiscoveryClientSuite.scala [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  private var etcdCluster: EtcdCluster = _
+  var engineServer: NoopTBinaryFrontendServer = _
+
+  private lazy val _connectString = etcdCluster.clientEndpoints().asScala.mkString(",")
+
+  override def getConnectString(): String = _connectString
+
+  val conf: KyuubiConf = {
+    KyuubiConf()
+      .set(HA_CLIENT_CLASS, classOf[EtcdDiscoveryClient].getName)
```
Comment (zwangsheng) [inline kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/etcd/EtcdDiscoveryClientSuite.scala:43]: How about we describe the option `EtcdDiscoveryClient ` in the `HA_CLIENT_CLASS` docs, after all, ETCD is not mentioned in the docs at present.

**Review by zwangsheng — APPROVED** [review]
LGTM

File: .rat-excludes [hunk 12]
```diff
@@ -44,6 +44,8 @@ build/scala-*/**
 **/semantic.min.css
 **/icon.png
 **/icon.min.css
+**/io.grpc.LoadBalancerProvider
+**/io.grpc.NameResolverProvider
```
Comment (yaooqinn) [inline .rat-excludes:48]: nit: remove them from rat-ex

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master
